server-utils/src/suspendable_stream.rs

```diff
@@ -0,0 +1,63 @@
+use std::time::{Duration, Instant};
+use tokio::timer::Delay;
+use std::io;
+use tokio::prelude::*;
+
+/// A wrapper type for tokio::Incoming that stops accepting new connections
+/// for a specified amount of time once `ulimit` is reached
+pub struct SuspendableStream<S> {
+	stream: S,
+	delay: Duration,
+	timeout: Option<Delay>,
+}
+
+impl<S> SuspendableStream<S> {
+	/// construct a new Suspendable stream, given tokio::Incoming
+	/// and the amount of time to pause for.
+	pub fn new(stream: S) -> Self {
+		SuspendableStream {
+			stream,
+			delay: Duration::from_millis(500),
+			timeout: None,
+		}
+	}
+}
+
+impl<S, I> Stream for SuspendableStream<S>
+	where S: Stream<Item=I, Error=io::Error>
+{
+	type Item = I;
+	type Error = ();
+
+	fn poll(&mut self) -> Result<Async<Option<Self::Item>>, ()> {
+		let mut recovered = None;
```

Comment:
Maybe call this `resumed` – we don't know if we've recovered at this point, we just know we're ready to try again, right?

Reply:
And that can just be an boolean, does not need to be an Option.